### PR TITLE
fix hparam wer_file never used in recipes

### PR DESCRIPTION
**Pull request by Adel-Moumen**
This PR aims at fixing an issue with `wer_file` hparam that is never used in our recipes. 

Indeed, in most of them, we are always overriding the value of `were_file`. For instance, see CommonVoice (https://github.com/speechbrain/speechbrain/blob/develop/recipes/CommonVoice/ASR/CTC/train_with_wav2vec.py#L384), LibriSpeech (https://github.com/speechbrain/speechbrain/blob/develop/recipes/LibriSpeech/ASR/CTC/train_with_wav2vec.py#L432), KSponSpeech (https://github.com/speechbrain/speechbrain/blob/5f471bae30b220d873db6fc0f6f3a6690b49c856/recipes/KsponSpeech/ASR/transformer/train.py#L506), and there are plenty remaining..... 

Credits goes to Yannick Estève for spotting this issue. 

## How I solved it? 

I introduce a new function `generate_wer_filename(wer_file: str, label: str) -> str` in `utils/data_utils.py` which basically generate the new `wer_file` based on the hparam `wer_file` and a label where the label can be any string. In most recipes, the new label is `test-clean`, `test-other` or `valid`. 

## Tests
I ran a `--test_only` recipe with Whisper and got the following wer files:
>ls train_whisper/1986/
dev-clean.csv  hyperparams.yaml  opt_librispeech_prepare.pkl  test-clean.csv  train-clean-100.csv  train.csv      train-other-500.csv    wer_test-clean.txt                      wer_test_if_its_working_test-other.txt
env.log        log.txt           save                         test-other.csv  train-clean-360.csv  train_log.txt  train_with_whisper.py  wer_test_if_its_working_test-clean.txt  wer_test-other.txt
 
I tested the changes with: 
```yaml
wer_file: !ref <output_folder>/wer.txt
``` 
and
```yaml
wer_file: !ref <output_folder>/wer_test_if_its_working.txt
```

### CODE REVIEW

**Comment by mravanelli** [comment]
Thank you @Adel-Moumen for this PR! I've conducted an initial review of the code and have a few suggestions for improvement:
1. Instead of introducing a new function for a simple operation like generating filenames, consider using a YAML approach. You could define the expected wer files (e.g., wer_test_file and wer_dev_file) directly in the YAML configuration and ensure they are appropriately utilized when needed.
2. I propose to have naming consistency across the recipes. We can use the following naming convention: "wer_test.txt" for testing and "wer_valid.txt" for validation. LibriSpeech could be an exception, as we have "wer_test_clean.txt" and "wer_test_others.txt". To fix: in the file recipes/LibriSpeech/ASR/transformer/hparams/train_hf_whisper.yaml, update the filename to be consistent: wer_test.txt.
3. I propose taking this opportunity for incorporating the "test_debug_checks" feature for all recipe tests related to ASR. You can refer to the link [https://github.com/speechbrain/speechbrain/blob/develop/tests/recipes/LibriSpeech.csv] for guidance. This addition will allow us to verify the correct creation of the wer file and other expected files. While we can gradually integrate the test_debug_checks for all recipe tests, I suggest prioritizing the ASR-related ones.

**Comment by Adel-Moumen** [comment]
I think you are right in your approach @mravanelli. I changed everything to be more YAML oriented. 

Could you please have a look and tell me if it's ok now? 

After that, I will add the name of the output wer in the test recipes. 

Note. I did not added the Switchboard recipe as I don't know what are the name of the train splits.

**Comment by Adel-Moumen** [comment]
Note: for LibriSpeech and Switchboard I introduced a new YAML variable called `output_wer_folder: !ref <output_folder>/`. The main reason is that we can't add in the YAML one variable for test-clean, one for test-other, etc... because the name of a test split is solely based on the name of the test file (e.g test-other.csv will be test-other). And this is an issue. Because when we are running the recipe test tool we are setting the test file to be `wer_ASR_train` which means, the name of the split will be `wer_ASR_train`, and thus, will not match a specific key e.g. `test-other_wer_file: ...` as it doesn't exist. 

So, to summarize, for the recipes which have only one test file, there's no changes, we are providing in the yaml the name of the output file while for recipes with multiple tests files, we are giving a folder.

**Comment by mravanelli** [comment]
I fixed a few small issues in the recipe tests. However, it looks like some librispeech recipe tests are still failing for reasons not connected to this PR. It is however important to investigate and maybe fix them directly here:
1.` ERROR: The variable test WER of file tests/download/LibriSpeech_ASR_seq2seq/train_log.txt (recipe full_inference_row_3) `violated the specified threshold (100.0 <=4.17). It looks like the problem here is that the checkpoint was not loaded correctly. The test was done on epoch 0 and thus it failed.
2. For this recipe test `recipes/LibriSpeech/ASR/transducer/hparams/conformer_transducer.yaml`, I got the following error in the checkpoints:
```
Traceback (most recent call last):
  File "/workspace/speechbrain_fixwer/recipes/LibriSpeech/ASR/transducer/train.py", line 524, in <module>
    asr_brain.evaluate(
  File "/workspace/speechbrain_fixwer/speechbrain/core.py", line 1550, in evaluate
    self.on_evaluate_start(max_key=max_key, min_key=min_key)
  File "/workspace/speechbrain_fixwer/recipes/LibriSpeech/ASR/transducer/train.py", line 299, in on_evaluate_start
    ckpt = sb.utils.checkpoints.average_checkpoints(
  File "/workspace/speechbrain_fixwer/speechbrain/utils/checkpoints.py", line 1201, in average_checkpoints
    return averager(parameter_iterator)
  File "/workspace/speechbrain_fixwer/speechbrain/utils/checkpoints.py", line 1107, in average_state_dicts
    raise ValueError("No state dicts to average.")
ValueError: No state dicts to average.
```
I'm sure this test was passing a few days ago. Did you change something relevant? Please let me know if you are able to replicate.

**Comment by Adel-Moumen** [comment]
> I fixed a few small issues in the recipe tests. However, it looks like some librispeech recipe tests are still failing for reasons not connected to this PR. It is however important to investigate and maybe fix them directly here: 1.`ERROR: The variable test WER of file tests/download/LibriSpeech_ASR_seq2seq/train_log.txt (recipe full_inference_row_3)`violated the specified threshold (100.0 <=4.17). It looks like the problem here is that the checkpoint was not loaded correctly. The test was done on epoch 0 and thus it failed. 2. For this recipe test `recipes/LibriSpeech/ASR/transducer/hparams/conformer_transducer.yaml`, I got the following error in the checkpoints:
> 
> ```
> Traceback (most recent call last):
>   File "/workspace/speechbrain_fixwer/recipes/LibriSpeech/ASR/transducer/train.py", line 524, in <module>
>     asr_brain.evaluate(
>   File "/workspace/speechbrain_fixwer/speechbrain/core.py", line 1550, in evaluate
>     self.on_evaluate_start(max_key=max_key, min_key=min_key)
>   File "/workspace/speechbrain_fixwer/recipes/LibriSpeech/ASR/transducer/train.py", line 299, in on_evaluate_start
>     ckpt = sb.utils.checkpoints.average_checkpoints(
>   File "/workspace/speechbrain_fixwer/speechbrain/utils/checkpoints.py", line 1201, in average_checkpoints
>     return averager(parameter_iterator)
>   File "/workspace/speechbrain_fixwer/speechbrain/utils/checkpoints.py", line 1107, in average_state_dicts
>     raise ValueError("No state dicts to average.")
> ValueError: No state dicts to average.
> ```
> 
> I'm sure this test was passing a few days ago. Did you change something relevant? Please let me know if you are able to replicate.

1. is fixed. 
2. I don't know why it is happening Im checking.

**Comment by Adel-Moumen** [comment]
Regarding 2. the issue seems to be resolved with the latests commits.

> (1/29) Running test for full_inference_row_2...
        download_file('https://www.openslr.org/resources/12/test-clean.tar.gz', 'tests/download/test-clean.tar.gz', unpack=True, dest_unpack='tests/download/LibriSpeech',write_permissions=True)
tests/download/test-clean.tar.gz exists. Skipping download
Extracting tests/download/test-clean.tar.gz to tests/download/LibriSpeech
        download_file('https://www.dropbox.com/sh/wy6ktfdxqt0dmou/AABY4ty3zy3xPef0zwRx949Oa?dl=1', 'tests/download/LibriSpeech_ASR_transformer.zip', unpack=True, dest_unpack='tests/download/LibriSpeech_ASR_transformer', write_permissions=True)
tests/download/LibriSpeech_ASR_transformer.zip exists. Skipping download
Extracting tests/download/LibriSpeech_ASR_transformer.zip to tests/download/LibriSpeech_ASR_transformer
                Running an Inference Test.
        ... 38.90s
(2/29) Running test for full_inference_row_3...
        download_file('https://www.openslr.org/resources/12/test-clean.tar.gz', 'tests/download/test-clean.tar.gz', unpack=True, dest_unpack='tests/download/LibriSpeech', write_permissions=True)
tests/download/test-clean.tar.gz exists. Skipping download
Extracting tests/download/test-clean.tar.gz to tests/download/LibriSpeech
        download_file('https://www.dropbox.com/sh/vs1os5mwh4necp0/AAAs7rMl4IO4N4s__BbVon6qa?dl=1', 'tests/download/LibriSpeech_ASR_seq2seq.zip', unpack=True, dest_unpack='tests/download/LibriSpeech_ASR_seq2seq', write_permissions=True)
tests/download/LibriSpeech_ASR_seq2seq.zip exists. Skipping download
Extracting tests/download/LibriSpeech_ASR_seq2seq.zip to tests/download/LibriSpeech_ASR_seq2seq
                Running an Inference Test.
        ... 20.93s
(3/29) Running test for LibriSpeech_row_2...
        ... 35.65s
(4/29) Running test for LibriSpeech_row_3...
        ... 24.86s
(5/29) Running test for LibriSpeech_row_4...
        ... 26.48s
(6/29) Running test for LibriSpeech_row_5...
        ... 35.55s
(7/29) Running test for LibriSpeech_row_6...
        ... 15.24s
(8/29) Running test for LibriSpeech_row_7...
        ... 46.22s
(9/29) Running test for LibriSpeech_row_8...
        ... 50.03s
(10/29) Running test for LibriSpeech_row_9...
        ... 25.32s
(11/29) Running test for LibriSpeech_row_10...
        ... 34.35s
(12/29) Running test for LibriSpeech_row_11...
        ... 37.87s
(13/29) Running test for LibriSpeech_row_12...
        ... 37.63s
(14/29) Running test for LibriSpeech_row_13...
        ... 36.67s
(15/29) Running test for LibriSpeech_row_14...
        ... 41.19s
(16/29) Running test for LibriSpeech_row_15...
        ... 36.13s
(17/29) Running test for LibriSpeech_row_18...
        ... 336.88s
(18/29) Running test for LibriSpeech_row_19...
        ... 175.63s
(19/29) Running test for LibriSpeech_row_20...
        ... 43.73s
(20/29) Running test for LibriSpeech_row_21...
        ... 9.73s
(21/29) Running test for LibriSpeech_row_22...
        ... 28.07s
(22/29) Running test for LibriSpeech_row_23...
        ... 16.65s
(23/29) Running test for LibriSpeech_row_28...
        ... 4.08s
(24/29) Running test for LibriSpeech_row_29...
        ... 3.94s
(25/29) Running test for LibriSpeech_row_30...
        ... 17.88s
(26/29) Running test for LibriSpeech_row_31...
        ... 16.24s
(27/29) Running test for LibriSpeech_row_32...
        ... 19.70s
(28/29) Running test for LibriSpeech_row_33...
        ... 49.69s
(29/29) Running test for LibriSpeech_row_34...
        ... 64.37s
TEST PASSED

**Comment by mravanelli** [comment]
Thank you @Adel-Moumen, all the recipe tests pass. I think it is now safe to merge this PR now

**Review by mravanelli — APPROVED** [review]
